### PR TITLE
`<flat_meow>`: Add tests for all `flat_map` constructors, deduction guides and then some

### PR DESCRIPTION
**Pull request by vmichal**
Changed functionality:
- Fix optimization of heterogeneous `flat_map::erase(OtherKey&&)` when `OtherKey` is not `key_type` .

Tests:
- Test that `flat_meow::insert`, `emplace_hint` and similar functions accepting hint correctly handle correct hint (insertion takes place as close as possible to given hint)
- Test all (I believe) constructors and deduction guides of `flat_map`
- Add death tests for all `_STL_ASSERT`s in both `<flat_map>` and `<flat_set>` - unsorted/duplicate elements in ranges and size mismatch of key and mapped container.
- Test all heterogeneous lookup and erase operations (related to previous work in #5994, see discussion in #5992 ).
- Extract common tests of container requirements and invariants from flat_set test and also apply them to flat_map objects.

Progress towards https://discord.com/channels/737189251069771789/1120873380615295027/1450952672982204468
> Also to exercise every deduction guide. [...] I want to make sure that every way we construct the container adaptors, we establish the invariants (so we actually-sort and actually-unique input when necessary).

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
I am reviewing this now and will resolve the trivial adjacent-add conflict.

File: tests/std/include/test_container_requirements.hpp [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+void assert_is_sorted_maybe_unique(const T& s) {
+    const auto val_comp = s.value_comp();
+    auto begin_it       = s.cbegin();
+    const auto end_it   = s.cend();
+
+    // internal check by the container itself
+    assert(s._Is_sorted_and_unique());
+
+    // external check observable by the user
+    assert(std::is_sorted(begin_it, end_it, val_comp));
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp:133]: Need `<algorithm>` for `std::is_sorted()`.

File: tests/std/include/test_container_requirements.hpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    if constexpr (ExpectedUnique) {
+        if (!s.empty()) {
+            for (auto prev_it = begin_it, it = prev_it + 1; it != end_it; ++prev_it, ++it) {
+                const bool val_comp_ok = val_comp(*prev_it, *it);
+                if constexpr (std::formattable<T, char>) {
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: Need `<format>` for `std::formattable`. `<print>` isn't guaranteed to drag this in.

File: tests/std/include/test_container_requirements.hpp [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    assert(empty.empty());
+    assert(non_empty == s);
+
+    assert(s.cbegin() <= s.cend());
+    assert(s.cbegin() < s.cend() || s.empty());
+
+    assert(m.begin() <= m.end());
+    assert(m.begin() < m.end() || m.empty());
+
+    assert(static_cast<typename T::size_type>(s.cend() - s.cbegin()) == s.size());
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: `static_cast` is a Down With Typename context.

File: tests/std/include/test_container_requirements.hpp [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    static_assert(std::same_as<std::const_iterator<const_iterator>, const_iterator>);
+    static_assert(std::is_convertible_v<iterator, const_iterator>);
+
+    // additionally:
+    static_assert(std::is_same_v<value_type, std::pair<key_type, mapped_type>>);
+}
+
+template <typename T>
+void assert_three_way_comparability() {
+    using value_type = typename T::value_type;
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: This is a Down With Typename context.

File: tests/std/include/test_container_requirements.hpp [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+#include <type_traits>
+#include <utility>
+
+// Extracted common functionality from flat_map and flat_set tests.
+// May be extended for other containers if needed
+
+template <class T>
+void assert_container_requirements(const T& s) {
+    T m = s;
+    assert(m == s);
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp:21]: Need `<cassert>` for `assert`.

File: tests/std/include/test_container_requirements.hpp [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    typename T::container_type;
+    requires std::is_nothrow_swappable_v<typename T::container_type>;
+};
+
+template <typename T>
+concept has_nothrow_swappable_containers =
+    set_has_nothrow_swappable_containers<T> || map_has_nothrow_swappable_containers<T>;
+
+template <typename T>
+struct is_nothrow_swappable {};
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: This shadows the Standard type trait and appears to be unused.

File: tests/std/include/test_container_requirements.hpp [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    static_assert(std::same_as<std::const_iterator<const_iterator>, const_iterator>);
+    static_assert(std::is_convertible_v<iterator, const_iterator>);
+
+    // additionally:
+    static_assert(std::is_same_v<key_type, value_type>);
+    static_assert(std::same_as<std::const_iterator<iterator>, iterator>);
+    static_assert(std::is_convertible_v<const_iterator, iterator>);
+    static_assert(std::is_same_v<decltype(*std::declval<T>().begin()), const typename T::value_type&>);
+    static_assert(std::is_same_v<std::remove_cvref_t<typename T::reference>, typename T::value_type>);
+    static_assert(std::is_same_v<typename T::const_reference, const typename T::value_type&>);
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: This function already has `using value_type     = T::value_type;` so it should take advantage of that for readability. And it should extract `reference` and `const_reference` while it's in the neighborhood.

File: tests/std/include/test_container_requirements.hpp [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    static_assert(std::is_same_v<decltype(s.size()), typename T::size_type>);
+    static_assert(std::is_same_v<decltype(s.max_size()), typename T::size_type>);
+    static_assert(std::is_same_v<decltype(*m.begin()), typename T::reference>
+                  || std::is_same_v<decltype(*m.begin()), typename T::const_reference>);
+    static_assert(std::is_same_v<decltype(*m.cbegin()), typename T::const_reference>);
+
+    T my_moved = std::move(m);
+    assert(!(my_moved != s));
+
+    T empty{};
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: This shadows `std::empty` and generally makes things hard to read.

File: tests/std/include/test_container_requirements.hpp [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    static_assert(std::is_same_v<decltype(T{}.rbegin()), typename T::reverse_iterator>);
+    static_assert(std::is_same_v<decltype(T{}.rend()), typename T::reverse_iterator>);
+    static_assert(std::is_same_v<decltype(s.rbegin()), typename T::const_reverse_iterator>);
+    static_assert(std::is_same_v<decltype(s.rend()), typename T::const_reverse_iterator>);
+    static_assert(std::is_same_v<decltype(s.crbegin()), typename T::const_reverse_iterator>);
+    static_assert(std::is_same_v<decltype(s.crend()), typename T::const_reverse_iterator>);
+    static_assert(std::is_convertible_v<typename T::reverse_iterator, typename T::const_reverse_iterator>);
+}
+
+template <typename T>
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: Although old tests are inconsistent (my fault), we try to consistently use `class` for template parameters in new tests.

File: tests/std/include/test_container_requirements.hpp [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    empty.swap(non_empty);
+    assert(non_empty.empty());
+    assert(empty == s);
+
+    std::swap(empty, non_empty);
+    assert(empty.empty());
+    assert(non_empty == s);
+
+    assert(s.cbegin() <= s.cend());
+    assert(s.cbegin() < s.cend() || s.empty());
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: This could be stricter.

File: tests/std/include/test_container_requirements.hpp [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        static_assert(noexcept(s.swap(s)) == is_noexcept);
+        static_assert(noexcept(std::ranges::swap(s, s)) == is_noexcept); // using ADL-swap
+        static_assert(noexcept(s.clear()));
+    }
+}
+
+template <bool ExpectedUnique, class T>
+void assert_is_sorted_maybe_unique(const T& s) {
+    const auto val_comp = s.value_comp();
+    auto begin_it       = s.cbegin();
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: `begin_it` could be const too.

File: tests/std/include/test_container_requirements.hpp [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+void assert_container_requirements(const T& s) {
+    T m = s;
+    assert(m == s);
+
+    static_assert(std::is_same_v<decltype(m = s), T&>);
+    static_assert(std::is_same_v<decltype(m = std::move(m)), T&>);
+    static_assert(std::is_same_v<decltype(m.begin()), typename T::iterator>);
+    static_assert(std::is_same_v<decltype(m.end()), typename T::iterator>);
+    static_assert(std::is_same_v<decltype(s.cbegin()), typename T::const_iterator>);
+    static_assert(std::is_same_v<decltype(m.cend()), typename T::const_iterator>);
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp:30]: This could be stricter, to test the const overloading.

File: tests/std/include/test_container_requirements.hpp [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+    static_assert(std::is_same_v<decltype(s.crend()), typename T::const_reverse_iterator>);
+    static_assert(std::is_convertible_v<typename T::reverse_iterator, typename T::const_reverse_iterator>);
+}
+
+template <typename T>
+concept map_has_nothrow_swappable_containers = requires {
+    typename T::key_container_type;
+    typename T::mapped_container_type;
+    requires std::is_nothrow_swappable_v<typename T::mapped_container_type>;
+    requires std::is_nothrow_swappable_v<typename T::key_container_type>;
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: Nitpick: These are out of order.

File: tests/std/include/test_container_requirements.hpp [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+    // internal check by the container itself
+    assert(s._Is_sorted_and_unique());
+
+    // external check observable by the user
+    assert(std::is_sorted(begin_it, end_it, val_comp));
+    if constexpr (ExpectedUnique) {
+        if (!s.empty()) {
+            for (auto prev_it = begin_it, it = prev_it + 1; it != end_it; ++prev_it, ++it) {
+                const bool val_comp_ok = val_comp(*prev_it, *it);
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: This is an `adjacent_find`.

File: tests/std/include/test_container_requirements.hpp [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+                if constexpr (std::formattable<T, char>) {
+                    if (!val_comp_ok) {
+                        std::println("Container {} is not sorted-maybe-unique", s);
```
Comment (StephanTLavavej) [inline tests/std/include/test_container_requirements.hpp]: We've asserted `is_sorted` and tested `if constexpr (ExpectedUnique)` so we can be more precise with our phrasing.

File: tests/std/tests/P1222R4_flat_set/test.cpp [hunk 16]
```diff
@@ -937,6 +933,42 @@ void test_erase_2() {
...
+};
+template <class Cont>
+concept can_erase_iterator_holder = requires(Cont c) {
+    { c.erase(holder<typename Cont::const_iterator>{c.begin()}) } -> same_as<typename Cont::iterator>;
+};
+// Note that std::less<T> is not transparent, std::less<> and ranges::less are
+
+namespace detail {
+    using C = flat_set<int>;
+    static_assert(same_as<C::iterator, decltype(std::declval<C>().erase(std::declval<C::iterator>()))>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P1222R4_flat_set/test.cpp]: This test is `using namespace std;` and there's no concern about ambiguity with `ranges` for `declval`.

File: tests/std/tests/P1222R4_flat_set/test.cpp [hunk 17]
```diff
@@ -265,19 +162,44 @@ void assert_all_requirements_and_equals(const T& s, const initializer_list<typen
...
-    assert_all_requirements_and_equals(flat_set<int, lt, C>(C{3, 7, 1, 85, 222, 1}), {1, 3, 7, 85, 222});
-    assert_all_requirements_and_equals(flat_multiset<int, lt, C>(C{3, 7, 1, 85, 7, 222, 1}), {1, 1, 3, 7, 7, 85, 222});
-    assert_all_requirements_and_equals(flat_set<int, gt, C>(C{1, 2, 3, 3}, gt()), {3, 2, 1});
-    assert_all_requirements_and_equals(flat_multiset<int, gt, C>(C{1, 1, 2, 3}, gt()), {3, 2, 1, 1});
-    assert_all_requirements_and_equals(flat_set<int, gt, C>(sorted_unique, C{30000, 200, 1}, gt()), {30000, 200, 1});
-    assert_all_requirements_and_equals(flat_multiset<int, gt, C>(sorted_equivalent, C{3, 3, -1}, gt()), {3, 3, -1});
+    {
+        // Test flat_set()
+        // and  flat_set(const key_compare&)
+        const std::less<int> comp;
```
Comment (StephanTLavavej) [inline tests/std/tests/P1222R4_flat_set/test.cpp]: This function has an `lt` alias.

File: tests/std/tests/P1222R4_flat_set/test.cpp [hunk 18]
```diff
@@ -265,19 +162,44 @@ void assert_all_requirements_and_equals(const T& s, const initializer_list<typen
...
+            assert_all_requirements_and_equals(fms, fms1, {});
+        }
+    }
+    {
+        // Test flat_set(const container)
+    }
+
+    assert_all_requirements_and_equals(flat_set(C{3, 7, 1, 85, 222, 1}), {1, 3, 7, 85, 222});
+    assert_all_requirements_and_equals(flat_multiset(C{3, 7, 1, 85, 7, 222, 1}), {1, 1, 3, 7, 7, 85, 222});
+    assert_all_requirements_and_equals(flat_set(C{1, 2, 3, 3}, gt()), {3, 2, 1});
```
Comment (StephanTLavavej) [inline tests/std/tests/P1222R4_flat_set/test.cpp]: When possible, we prefer to use braces to construct temporaries, as it avoids looking like a function call. (Not always possible in generic code.)

File: tests/std/tests/P1222R4_flat_set/test.cpp [hunk 19]
```diff
@@ -493,6 +415,65 @@ void test_insert_2() {
...
 
+// Test that hint to emplace/insert is respected, when possible; check returned iterator
+template <class C>
+void test_insert_hint_is_respected() {
+    using lt = std::less<int>;
+
+    {
+        flat_multiset<int, lt, C> a{-1, -1, 1, 1};
+        bool problem_seen                      = false;
+        auto const assert_inserted_at_position = [&a, &problem_seen](
```
Comment (StephanTLavavej) [inline tests/std/tests/P1222R4_flat_set/test.cpp]: We use west const.

File: tests/std/tests/P1222R4_flat_set/test.cpp [hunk 20]
```diff
@@ -937,6 +933,42 @@ void test_erase_2() {
...
+    static_assert(same_as<C::iterator, decltype(std::declval<C>().erase(std::declval<holder<C::const_iterator>>()))>);
+    static_assert(same_as<C::iterator,
+        decltype(std::declval<C>().erase(std::declval<C::iterator>(), std::declval<C::iterator>()))>);
+    static_assert(same_as<C::size_type, decltype(std::declval<C>().erase(std::declval<holder<int>>()))>);
+    static_assert(same_as<C::size_type, decltype(std::declval<C>().erase(std::declval<int>()))>);
+
+    // erase key_type
+    static_assert(can_erase_key<flat_set<int, _STD less<int>>, int>);
+    static_assert(can_erase_key<flat_set<int, _STD less<>>, int>);
+    static_assert(can_erase_key<flat_set<int, _RANGES less>, int>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P1222R4_flat_set/test.cpp]: This is using our internal namespace macros which we avoid (for things like `_CONSTEXPR20` it can be acceptable but we know how to spell `std::`).

File: tests/std/tests/P0429R9_flat_map/test.cpp [hunk 21]
```diff
@@ -72,6 +57,17 @@ bool check_value_content(const T& obj, const typename T::mapped_container_type&
...
+template <IsFlatMap T>
+bool assert_check_content(const T& obj, const type_identity_t<T>& expected) {
+    assert_all_requirements(obj);
+    if (!ranges::equal(obj, expected)) {
+        println(stderr, "Unexpected content!\nExpected {}", expected);
+        println(stderr, "Actual {}", obj);
+        return false;
+    }
+    return true;
+}
```
Comment (StephanTLavavej) [inline tests/std/tests/P0429R9_flat_map/test.cpp:70]: It is confusing for a function to be named `assert_meow` when it doesn't assert its own condition. It should be named `check_meow` like the functions above.

File: tests/std/tests/P0429R9_flat_map/test.cpp [hunk 22]
```diff
@@ -152,6 +148,32 @@ class MyAllocator : public allocator<T> {
...
+namespace std {
+    template <class T, class U, size_t N>
+    struct tuple_element<N, almost_pair<T, U>> {
+        using type = conditional_t<N == 1, T, U>;
+    };
+
+    // template<class T, class U>
+    // struct tuple_size<almost_pair<T, U>> {
+    //     using type = conditional_t<N == 1, T, U>;
+    // };
```
Comment (StephanTLavavej) [inline tests/std/tests/P0429R9_flat_map/test.cpp]: We shouldn't have commented-out code and this makes no sense.

File: tests/std/tests/P0429R9_flat_map/test.cpp [hunk 23]
```diff
@@ -200,141 +222,525 @@ struct TransparentPackagedCompare : PackagedCompare<T> {
...
-        flat_multimap fmmap(keys, vals);
-        assert(check_requirements(fmmap));
-        assert(check_key_content(fmmap, {0, 1, 2, 2, 3, 4}));
-        assert(check_value_content(fmmap, {44, 2324, 635462, 7, 433, 5},
-            {
-                {0, 1, subrange_type::equal},
-                {2, 3, subrange_type::permutation},
-                {4, 5, subrange_type::equal},
-            }));
+        // Test flat_map(key_cont, mapped_cont, comp = key_comp()())
```
Comment (StephanTLavavej) [inline tests/std/tests/P0429R9_flat_map/test.cpp]: Damaged comment.

File: tests/std/tests/P0429R9_flat_map/test.cpp [hunk 24]
```diff
@@ -687,6 +1092,29 @@ void test_comparison() {
     }
 }
 
+template <template <class...> class Map>
+void test_map_operations_transparent() {
+    struct shouldnt_convert {
+        int key;
+        /* implicit */ [[noreturn]] operator int() const {
+            abort();
```
Comment (StephanTLavavej) [inline tests/std/tests/P0429R9_flat_map/test.cpp:1095]: Should include `<cstdlib>` for `abort()`.

**Comment by StephanTLavavej** [comment]
:green_heart:
===
